theano/gof/destroyhandler.py

```diff
@@ -114,43 +114,11 @@ def _contains_cycle(fgraph, orderings):
 
     # Pass through all the nodes to build visitable, parent_count, and
     # node_to_children
-    for var in fgraph.variables:
-
-        # this is faster than calling get_parents
-        owner = var.owner
-        # variables don't appear in orderings, so we don't need to worry
-        # about that here
-        if owner:
-            # insert node in node_to_children[r]
-            # (if r is not already in node_to_children,
-            # intialize it to [])
-            node_to_children.setdefault(owner, []).append(var)
-            parent_counts[var] = 1
-        else:
-            visitable.append(var)
-            parent_counts[var] = 0
-
-    for a_n in fgraph.apply_nodes:
-        parents = list(a_n.inputs)
-        # This is faster than conditionally extending
-        # IG: I tried using a shared empty_list = [] constructed
-        # outside of the for loop to avoid constructing multiple
-        # lists, but this was not any faster.
-        parents.extend(orderings.get(a_n, []))
-
-        if parents:
-            for parent in parents:
-                # insert node in node_to_children[r]
-                # (if r is not already in node_to_children,
-                # intialize it to [])
-                node_to_children.setdefault(parent, []).append(a_n)
-            parent_counts[a_n] = len(parents)
-        else:
-            # an Apply with no inputs would be a weird case, but I'm
-            # not sure we forbid it
-            visitable.append(a_n)
-            parent_counts[a_n] = 0
-
+    try:
+        graph.io_toposort(list(fgraph.variables), list(fgraph.apply_nodes))
+    except ValueError as ve:
+        if "cycles" in ve.args[0]:
+            return True
```

Comment:
This is just a branch. not a PR.

On Mon, Apr 4, 2016 at 1:06 PM, Ramana Subramanyam <notifications@github.com
> wrote:

> In theano/gof/destroyhandler.py
> <https://github.com/Theano/Theano/pull/4334#discussion_r58438924>:
>
> > -                # insert node in node_to_children[r]
> > -                # (if r is not already in node_to_children,
> > -                # intialize it to [])
> > -                node_to_children.setdefault(parent, []).append(a_n)
> > -            parent_counts[a_n] = len(parents)
> > -        else:
> > -            # an Apply with no inputs would be a weird case, but I'm
> > -            # not sure we forbid it
> > -            visitable.append(a_n)
> > -            parent_counts[a_n] = 0
> > -
> > +    try:
> > +        graph.io_toposort(list(fgraph.variables), list(fgraph.apply_nodes))
> > +    except ValueError as ve:
> > +        if "cycles" in ve.args[0]:
> > +            return True
>
> You've a PR opened ? I'll continue from there, in that case..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/4334/files/5dec82e8bb2d3ba0bdb2d55ce6ba045ed61dc41f#r58438924>
>